.github/workflows/publish.yml

```diff
@@ -0,0 +1,106 @@
+# Based on
+# https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/
+
+name: Publish Python distribution to PyPI
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
```

Comment:
Should this workflow really run on every PR? It seems unlikely to break and basically duplicates the normal test workflow. Also there is a slight risk that an error in this workflow might accidentally publish a package, and running it less frequently reduces that risk.

To me, it would make more sense to just run this workflow on commits to main.

Reply:
I wanted it to run on *this* PR so it gets tested :).

But agree we don't have to run it on every PR.

Reply:
Our existing workflow to test our packaging is run on every PR:

https://github.com/python/typing_extensions/blob/479ac33a0c97ea91c8177f6b6887ff7a9fb9970d/.github/workflows/package.yml#L1-L8

Relatedly, should that existing workflow be deleted as part of this PR? Or is it still relevant?

Reply:
Oh right, I saw that last night. I now removed package.yml and given that the typing-extensions test suite is extremely fast, it seems fine to keep running it on every PR.

Reply:
I still feel uncomfortable running this workflow on PRs for the reasons outlined above. I'd prefer to just run the test workflow and only run this workflow on the final merge to main.

Reply:
But then on the other hand, if we don't run this workflow on PRs, it increases the risk that packaging bugs like https://github.com/python/typing_extensions/issues/61 are introduced in PRs.

I still feel like I'd prefer to do the packaging _tests_ on every PR for that reason, given they only take 7s from start to finish. Maybe a compromise could be to put the `publish-to-pypi` job in a workflow by itself, and have that workflow only run on `push` events?

Reply:
> I still feel uncomfortable running this workflow on PRs for the reasons outlined above. I'd prefer to just run the test workflow and only run this workflow on the final merge to main.

To clarify: I don't mean this comment as a blocker! Overall I'm +1 on this PR. If you think it's better to do it this way, please do.

Reply:
The current tests in this PR wouldn't have caught #61, since they rely only on the installed version of typing_extensions.py and use the test helper files from the repo. I suppose we could add a step that manually unzips the sdist and runs tests on the files within it.

I think Sebastian's hesitations align with the recommendations in https://docs.pypi.org/trusted-publishers/security-model/ so I'll refactor with that in mind. I'm thinking:

- One workflow that runs on all PRs that builds the sdist and wheel, tests them, and uploads them as GH Actions artifacts.
- One workflow that runs only on releases that downloads the artifacts and uploads them to PyPI.

Once we have this merged, I plan to make an alpha release (I think it'd be 4.11a1) just to test the workflow. Once it works well and all of Alex's PRs are merged, we can release an RC for the next feature release.

Reply:
For mypy I set up https://github.com/mgedmin/check-manifest that catches "this is or isn't in sdist" issues

Reply:
I think running in separate workflows is not practical, since actions/download-artifact doesn't provide a simple way to share the artifacts (https://github.com/actions/download-artifact?tab=readme-ov-file#download-artifacts-from-other-workflow-runs-or-repositories - where would I get the run id?). Other repos I checked (pip-audit, hatch) also use a single workflow file.

However, the upload step is run in a separate environment, which I configured to require approval from one of the people on this PR. I think that should address the concerns about running this on PRs.

Reply:
> I think running in separate workflows is not practical, since actions/download-artifact doesn't provide a simple way to share the artifacts ([actions/download-artifact#download-artifacts-from-other-workflow-runs-or-repositories](https://github.com/actions/download-artifact?tab=readme-ov-file#download-artifacts-from-other-workflow-runs-or-repositories) - where would I get the run id?). Other repos I checked (pip-audit, hatch) also use a single workflow file.

Do we need to share the artifact -- couldn't we just do a fresh build in the separate workflow? I think it's pretty quick, no?

Reply:
We could but then we wouldn't be testing the exact artifact we're releasing. Sure, they *should* be the same, but what if they're not?

Reply:
Oh I see your point. Right, so we'd need to duplicate all the test jobs in the separate workflow. Okay, this seems reasonable then :-)